src/tarantool_url.c

```diff
@@ -0,0 +1,177 @@
+#include "php_tarantool.h"
+
+#include "tarantool_url.h"
+
+#include "third_party/uri.h"
+
+static const char *
+url_part_copy(const char *part, size_t part_len, int persistent) {
+	if (part == NULL || part_len == 0)
+		return NULL;
+	char *rv = pecalloc(sizeof(char), part_len + 1, persistent);
+	rv[part_len] = '\0';
+	memcpy(rv, part, part_len);
+	return (const char *)rv;
+}
+
+struct tarantool_url *
+tarantool_url_parse(const char *url, int persistent) {
+	struct uri parsed; memset(&parsed, 0, sizeof(struct uri));
+
+	if (uri_parse(&parsed, url) == -1)
+		return NULL;
+
+	/* Normalization of URI */
+	enum tarantool_url_type tp = TARANTOOL_URL_NONE;
+	const char *user = NULL,
+		   *pass = NULL,
+		   *host = NULL,
+		   *port = NULL,
+		   *path = NULL;
+	size_t user_len = 0,
+	       pass_len = 0,
+	       host_len = 0,
+	       port_len = 0,
+	       path_len = 0;
+
+	if ((parsed.host != NULL && strncmp(parsed.host, "unix/", sizeof("unix/") - 1) == 0) ||
+	    (parsed.scheme == NULL || strncmp(parsed.scheme, "unix", sizeof("unix") - 1) == 0)) {
+		if (parsed.query != NULL || parsed.fragment != NULL)
+			return NULL;
+
+		if (parsed.service_len == 0) parsed.service = NULL;
+		if (parsed.path_len == 0)    parsed.path    = NULL;
+
+		/* If both are set or not then throw error */
+		if ((parsed.service == NULL) == (parsed.path == NULL))
+			return NULL;
+
+		path     = parsed.service ? parsed.service     : parsed.path;
+		path_len = parsed.service ? parsed.service_len : parsed.path_len;
```

Comment:
It seems that `service` is not the same as `path`. Consider test/unit/uri.c in tarantool repository:

```
  6 test(const char *s, const char *scheme, const char *login, const char *password,           
  7      const char *host, const char *service, const char *path,
  8      const char *query, const char *fragment, int host_hint)
  9 {  

 82         test("host:service/path1/path2/path3", NULL, NULL, NULL, "host",                   
 83              "service", "/path1/path2/path3", NULL, NULL, 0);
```

Reply:
I understand it with the following test case from the same file:

```
202         test("unix/:/path1/path2/path3:/p1/p2?q1=v1&q2=v2#fragment",                       
203              NULL, NULL, NULL, "unix/", "/path1/path2/path3", "/p1/p2",                    
204              "q1=v1&q2=v2", "fragment", 3);  
```

`service` is path to a unix socket file and `path` is some logical (say, http) path.

But this is only for `unix/:/path` URIs. When we use `unix:///path` the path is placed in `path`.

So the implementation is correct.